### ISSUE TITLE
Basic Support for Annotated Image Formats

### ISSUE THREAD

**Issue by banesullivan** [ResonantGeoData/ResonantGeoData#88]
**Task:** Add support for annotated image formats [KWCOCO](https://gitlab.kitware.com/computer-vision/kwcoco) and JOIA (JOIA is not public)

I've decided to group these two formats together as their general structures are very similar and they will both require us to add a few new metadata fields and models to the `geodata` app. Also, we will not be supporting either format in full, so it makes sense to me to bundle them based on their overlapping concepts. By "not supported in full", I mean simplifying the fields we extract from either format and abstracting similar concepts across the two into our models as our goal is primarily to be able to ingest these data rather than to be able to output any imagery in our database to those formats.

Initial work is done in #91 

Implementing support for these formats will require:

- [x] Creating a new `Annotation` model that has a relationship to the `RasterEntry` model
    - [x] this will have a name/caption text field
    - [x] this will have a bounding box polygon
    - [x] a feature GeoJSON for label shapes (points/polygons/etc)
    - [x] this will have an optional "annotator" field that lists the name/entity of what made the annotation - this could be an algorithm, so this will just be a text field
    - [x] A general text field for notes
    - [x] JOIA and KWCOCO have "labels" - I think it makes sense just to add this as an optional text field.
    - [ ] KWCOCO annotations are defined in the image pixel space, so we will have to remap those annotations to world coordinates. - this will be dealt with when addressing #98 
- [x] Implementing a few more fields on the `RasterEntry` and `BandMeta` models:
    - [x] Add a checksum field for raster files - JOIA provides this, but should we compute a checksum on rasters by default?
    - [x] Add optional polygon field to `RasterEntry` for outlining non-null pixels
    - ~Perhaps add an optional version number?~
- [x] Possibly add a new `DerivedRaster` model that points to a base `RasterEntry` parent
    - JOIA has this concept of derived images that point to a parent image. This is useful in ML workflows where a processing routine alters the original image then annotates that image. I'm not sure what would be the best way to implement this model or if we should. Perhaps we just use the standard `RasterEntry` model for this and not track the parent image?
- [ ] Add basic proof of concept import/export routines for both formats.
    - waiting on #98 before starting this work
    - import KWCOCO with image annotations
    - wait for JOIA devs to decouple code from DB before making import script
    - add basic export routines for image data only (no support for annotations yet) for both formats

Some notes:

- The KWCOCO format is not geospatial in nature, so we will have to coerce it to be geospatial. There is no spatial reference in the KWCOCO format, so we will have to extract that from the raw image with GDAL and throw an error if the image does not have a spatial reference. 
- We will have to write and maintain a custom reader routine for the JOIA format unless the JOIA team can figure out how to separate their Marshmellow-sqlalchemy schema from needing to access the database.
- There are many additional features both the KWCOCO and JOIA formats add, but I think this base support of "annotated images" will give us a lot of flexibility and be able to generally support those datasets.

**Comment by wphicks** [comment]
A few quick thoughts based on questions here:

- In an ML context, labels may have more structure to them than just a standard text output. In particular, it is frequently useful to have some sort of hierarchy of labels (e.g. Object is an animal AND a mammal AND a cat). Of course that could be supported in the logic of whatever client interfaces with your system, so up to you whether you want any additional label structure to be  supported in your database.
- In work that uses JOIA, data checksums are primarily useful for disambiguating datasets (i.e. I can guarantee that the exact same pixels that appear in dataset A are not included in dataset B). Not sure how crucial it is for you, but I wouldn't be opposed to having it.
- Basically same story on the image "parent stuff" as for checksums.
- A custom JOIA reader should definitely not be necessary. If that's an urgent need, we can handle decoupling it from a Marshmallow-SqlAlchemy context.

**Comment by banesullivan** [comment]
> First point:

I currently do not foresee us implementing this within the geodata models, and I am thinking it makes sense to let the client interfacing with the data catalog handle those hierarchies and interpretation. If we step more towards full support for annotation/label hierarchies and ML-specific features, then we may end up with a concerning amount of feature bloat for the majority of our users who won't be leveraging any of that while never fully satisfying the needs of ML users.

I am thinking that our annotation model will support a `name/caption` text field and the client making searches would know all the subtypes and use that knowledge to perform a search with the API. E.g. some aerial imagery was classified with the hierarchy of "infrastructure" -> "road" -> "bridge", in the data catalog, we register a label as "bridge" because that's what would be listed in the JOIA format. If the client needs all rasters with the "infrastructure" label, then they will have to know to do a union with results for "road" and "bridge" as that label interpretation should be done by the client, IMO.

One workaround to support this could be a clear format storing the hierarchy of those names/captions. E.g. `infrastructure:road:bridge` for the "bridge" annotation. Then we can search across each - but still, the client importing the data to the GeoData catalog would have to properly set those labels as that heirarchy schema wouldn't be reflected in the JOIA or KWCOCO formats to my understanding.

> second and third points:

We can certainly add an optional `checksum` text field and a flag for our ETL routine to compute that checksum if needed. Otherwise, I may not include it in the first round as we currently do not read any data from imagery files and this could add unneeded overhead.

> A custom JOIA reader should definitely not be necessary. If that's an urgent need, we can handle decoupling it from a Marshmallow-SqlAlchemy context.

It would be really awesome if you all could jump on this for me! I am currently working on the model implementation and will start with KWCOCO because we can use it's Python package for the IO. Would you all be able to make the JOIA python package public? If not, we may have to put the JOIA ETL routine in the JOIA module and have a way to register it with the Geodata app if installed.

**Comment by banesullivan** [comment]
> data checksums are primarily useful for disambiguating datasets (i.e. I can guarantee that the exact same pixels that appear in dataset A are not included in dataset B). Not sure how crucial it is for you, but I wouldn't be opposed to having it.

Currently, we do not perform any such check on data added to the catalog. At the moment, there isn't really a need for us to do that because we do not actually store any data itself, so there isn't a high cost for us to store duplicate entries (its just meta data) - the cost of computing checksums on every new entry might be higher the cost of allowing duplicate entries.

With our current implementation - we extract the metadata once and keep track of the original data's location. We have no way of knowing whether those data have changed or moved so having a checksum could be very useful for the catalog. Thoughts, @manthey?

Also for formats like JOIA/KWCOCO, should we also be checksum-ing the JSON header files? In theory those could change and be out of data with the catalog too

**Comment by manthey** [comment]
Currently, we are using either minio or S3 as the object store for the dataset.  Ideally, we would also allow data from federated sources.  In S3, files are stored uniquely, and we can be reasonably certain that if a file changes, some metadata about it will also change.  However, both in federated sources and in minio running where files are in a shared file system, files can change without informing the server.

I think we will want to compute checksums on any data we use and store these checksums alongside the filefield.  We could expose a "verify files" endpoint that would recheck checksums, and, if changed, would perform some appropriate action.  This could be done automatically, but, since it has a distinct cost in data access and cpu, that should probably be defered until we determine what is needed.

It is worth figuring out what minio does if it has a shared data folder and files are added or removed (besides being changed).
